QueryCiv3/Sav.cs

```diff
@@ -1,125 +1,371 @@
-        public ContItem[] Cont
-        { get {
-            int ContCount = Wrld.ContinentCount;
-            List<ContItem> LeadList = new List<ContItem>();
-            for(int i=0; i< ContCount; i++)
-            {
-                int LeadOffset = Sav.SectionOffset("CONT", i+1);
-                LeadList.Add(new ContItem(this, LeadOffset));
-            }
-            return LeadList.ToArray();
-        }}
-        public LeaderItem[] Lead
-        { get {
-            int LeadCount = 32;
-            List<LeaderItem> LeadList = new List<LeaderItem>();
-            for(int i=0; i< LeadCount; i++)
-            {
-                int LeadOffset = Sav.SectionOffset("LEAD", i+1);
-                LeadList.Add(new LeaderItem(this, LeadOffset));
-            }
-            return LeadList.ToArray();
-        }}
-        public CityItem[] City
-        { get {
-            // Since "CITY" can sometimes appear in dirty data, let's find the first
-            //   "CITY" with 0x00000088 after it
-            int CityOffset = 0;
-            for(int i=0, chk=0; chk!=0x88; i++)
-            {
-                CityOffset = Sav.SectionOffset("CITY", i+1);
-                chk = Sav.ReadInt32(CityOffset);
+
+        public unsafe void CopyArray<T>(ref T[] data, int length) where T : unmanaged
+        {
+            data = new T[length];
+            int dataLength = length * sizeof(T);
+
+            fixed (void* destPtr = data) {
+                Buffer.MemoryCopy(scan, destPtr, dataLength, dataLength);
             }
-            int CityCount = 1; // TEMP HACK // Game.CityCount;
+            scan += dataLength;
+        }
 
-            CityItem[] CityList = new CityItem[CityCount];
-            for(int i=0; i< CityCount; i++)
+        public unsafe void Load(byte[] savBytes)
+        {
+            Sav = new Civ3File(savBytes);
+            // Load in any biq sections contained in Sav file, overwriting existing biq sections:
+            int BiqSectionLength = Sav.ReadInt32(38);
+            Bic.Load(Sav.GetBytes(BIQ_SECTION_START, BiqSectionLength));
+
+            fixed (byte* bytePtr = savBytes)
             {
-                CityList[i] = new CityItem(this, CityOffset);
-                // incomplete; still need to account for bytes after Ctzn and list w/count after buildings stats
-                CityOffset += 0x228 + (CityList[i].CitizenCount * 300) + (Bic.Bldg.Length * 8) + 0x30;
+                int* header;
+                scan = bytePtr + BIQ_SECTION_START + BiqSectionLength;
+                byte* end = bytePtr + savBytes.Length;
+
+                while (scan < end) {
+                    header = (int*)scan;
+
+                    // Every header in civ 3 files is exactly 4-chars long, which means they can be represented as 32-bit integers instead of strings
+                    // Switching off of these hex values is substantially faster than string switching, but comes at the expense of readability
```

Comment:
"is substantially faster" - How much faster is this, really?

From a conceptual standpoint, it makes sense that int comparisons are quicker than string comparisons.  And if we were iterating through every single byte checking, "does this match?" - a naive approach I once used when searching for SAV structures in memory, knowing what I was looking for but not knowing where it was - I'd expect it to be noticeable.

But here, it looks like we're making a couple dozen comparisons based on the number of sections?  If so, I'm skeptical that it would add up to much at all, relative to all the other work that has to be done.

It is well-documented with the section headers right there, and this is not likely to be a highly-modified piece of code, so it's fine to leave it as-is.  It just matched my "micro-optimization that's probably undetectable in real life" sensor.